keps/sig-api-machinery/4988-serve-pagination-from-cache/README.md

```diff
@@ -0,0 +1,395 @@
+- Ignoring the limit parameter defeats the purpose of pagination,
+  potentially returning massive responses that can overwhelm clients.
+
+This situation leads to unexpected behavior for users,
+where enabling pagination can degrade performance instead of improving it.
+
+### Goals
+
+- Improve the performance of paginated list requests to be comparable to non-paginated lists.
+- Make pagination behavior predictable and consistent.
+- Reduce load on etcd caused by paginated requests.
+
+### Non-Goals
+
+- Serve `resourceVersion="N"` request from watch cache
+- Support indices when paginating.
+- Eliminate all paginated list requests to etcd.
+
+## Proposal
+
+Leveraging the recent rewrite of the watchcache storage layer to use a B-tree
+(https://github.com/kubernetes/kubernetes/pull/126754), we propose to utilize
+B-tree snapshots to serve paginated lists.
+
+Mechanism:
+1. **Snapshot Creation:** When a paginated list request (with a limit parameter
+   set) is received, the API server will create a snapshot using the efficient
+   [Clone()](https://pkg.go.dev/github.com/google/btree#BTree.Clone) method.
+   This clone is a lazy copy, only duplicating necessary nodes, resulting in
+   minimal overhead.
+2. **Snapshot Storage:** The snapshot will be stored in memory, keyed by
+   resourceVersion.
+3. **Serving Subsequent Pages:** When a subsequent request with a continue token
+   arrives, the API server will:
+   - Extract the resourceVersion from the continue token.
+   - Retrieve the corresponding snapshot using the resourceVersion as the key.
+   - Use the snapshot to serve the requested page, or pass the request to etcd is snapshot is not available.
+4. **Snapshot Cleanup:** Snapshots will be subject to a Time-To-Live (TTL)
+   mechanism. We will reuse the existing watch event cleanup logic, which has a
+   75s TTL. This ensures that snapshots don't accumulate indefinitely.
+
+As still some pagination requests will be delegated to etcd, we will monitor the
+success rate by measuring the pagination cache hit vs miss ratio.
+
+Consideration: Should we start respecting the limit parameter?
+
+### Risks and Mitigations
+
+#### Client setting limit while not supporting pagination
+
+#### Memory overhead
+
+No, B-tree only store pointers the actual objects, not the object themselves.
+The objects are already cached to serve watch, so it should only add a small
+overhead for the B-tree structure itself, which is negligible compared to the
+size of the cached objects.
+
+#### Pagination request hitting another apiserver
+
+When connecting directly expectation is that a client will stay connected to the
+same apiserver. What happens when apiserver is behind loadbalancer?
+
+For setups with L4 loadbalancer apiserver can be configured with Goaway, which
+requests client reconnects periodically, however per request probability should
+be configured around 0.1%.
+
+For L7 loadbalancer the default algorithm usually is round-robin. For most LBs
```

Comment:
Trying to better understand this, is the worst case here as follows (assume 3 API Servers A, B, C):
1. Client hits API Server A: assuming the rv is cached, snapshot is created on receiving a LIST request with a limit parameter set
2. Client hits API Server B: no snapshot present, we delegate to etcd
3. Client hits API Server C: no snapshot present, we delegate to etcd 

so the performance degenerates to the current situation without cached pagination?

Reply:
This ^ is assuming all are on a version that has support for pagination from the cache. If there is one server that is on a minor version which does not have support, my understanding is that that would again be delegated to etcd

Reply:
Right, there is no regression, assuming that:
* We will delegate continue requests, that we don't have cached responses for, to etcd.
* We will not change how API server doesn't respect limit for RV="0".